### ISSUE TITLE
client go-generator should allow adding custom headers to outgoing requests

### ISSUE THREAD

**Issue by evantorrie** [ardielle/ardielle-tools#24]
Currently, the generator for client.go writes code that initializes outgoing HTTP headers to an empty map, and adds only what it needs internally for its specific purposes (e.g. Credentials header/value and If-None-Match ETag type headers)

It would be good to allow custom headers to be added to each client on a per-request basis.

**Comment by boynton** [comment]
This would be a worthwhile enhancement.

**Comment by boynton** [comment]
Do we want both custom "sticky" headers like the current credentials, as well as per-request custom headers? The latter would most easily be done by modifying the API to take a map, which would create incompatibilities. The sticky headers would be trivial to add

**Comment by evantorrie** [comment]
In the case I'm thinking about, we would be ok with sticky credentials - so long as there was some way to "update" them (e.g. to refresh a header value).  Even if that meant calling "AddHeaders()" again, that would probably be ok.

**Comment by boynton** [comment]
Would probably need to add API to SetHeader instead of AddHeader (which will accumulate a list of values). Sticky headers is easy enough that I can see doing that now, per-request headers may require some API work, not quite ready to do that.

**Comment by gotwarlost** [comment]
Why couldn't you do all this in a custom transport? We already do user-agent injection and other stuff like this using a custom transport.

**Comment by boynton** [comment]
Good point. The other reason I've dragged my feet is that headers are technically part of the API, and I'd prefer API details to be explicit in the RDL itself (an approach which also allows adding whatever headers you want in a more formal way).
